### ISSUE TITLE
[naming-convention] conflict when i use "filter"

### ISSUE THREAD

**Issue by FelixLgr** [typescript-eslint/typescript-eslint#4900]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "variable",
        "types": [
          "boolean"
        ],
        "format": [
          "PascalCase"
        ],
        "prefix": [
          "is",
          "should",
          "has",
          "can",
          "did",
          "will"
        ]
      },
      {
        "selector": "variable",
        "format": [
          "camelCase",
          "PascalCase",
          "UPPER_CASE"
        ],
        "filter": {
          "regex": "^_t$",
          "match": false
        }
      }
]
  }
}
```

```TS
const foo: boolean = true
const _t: string = "toto"
```

**Expected Result**
```TS
const foo: boolean = true //<-- Variable name `toto` must have one of the following prefixes: is, should, has, can, did, will
const _t: string = "toto" //<-- valid
```

**Actual Result**
```TS
const foo: boolean = true //<-- valid
const _t: string = "toto" //<-- valid
```

**Additional Info**

When my config is :
```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "variable",
        "types": [
          "boolean"
        ],
        "format": [
          "PascalCase"
        ],
        "prefix": [
          "is",
          "should",
          "has",
          "can",
          "did",
          "will"
        ]
      },
      {
        "selector": "variable",
        "format": [
          "camelCase",
          "PascalCase",
          "UPPER_CASE"
        ]
      }
]
  }
}
```
I get :

```TS
const foo: boolean = true //<-- Variable name `toto` must have one of the following prefixes: is, should, has, can, did, will
const _t: string = "toto" //<-- Variable name `_t` must match one of the following formats: camelCase, PascalCase, UPPER_CASE
```
I have just remove filter in the `variable` selector the filter param

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `6.8.0` |

**Comment by bradzacher** [comment]
```
        "filter": {
          "regex": "^_t$",
          "match": false
        }
```

You have `match: false`. This means that it will match every single name, ***except for*** `_t`.

Your other selector only matches `boolean` variables.

Because the string variable `_t` matches zero selectors - it is ignored entirely by the rule.

**Comment by FelixLgr** [comment]
> ```
>         "filter": {
>           "regex": "^_t$",
>           "match": false
>         }
> ```
> 
> You have `match: false`. This means that it will match every single name, ***except for*** `_t`.
> 
> Your other selector only matches `boolean` variables.
> 
> Because the string variable `_t` matches zero selectors - it is ignored entirely by the rule.

Thanks for your answer but why when I add the filter to the second rule, the `boolean` variable can be named foo without any prefix (is, should, has, ...) for me it is not logical.

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/naming-convention/#how-does-the-rule-automatically-order-selectors
